registry/quilt_server/views.py

```diff
@@ -2307,6 +2316,174 @@ def _comment_dict(comment):
+
+@app.route('/api/roles', methods=['POST'])
+@api(enabled=ENABLE_USER_ENDPOINTS, require_admin=True, schema=ROLE_DETAILS_SCHEMA)
+@as_json
+def add_role():
+    """
+    The body of the request should contain a JSON object.
+    There is one required parameter:
+        name(string): name of the role to operate on
+
+    There is an optional paramters:
+        arn(string): ARN of the IAM role associated with the Quilt role.
+
+    To create a role, you must provide an unused name and an arn.
+    """
+    data = request.get_json()
+    role_name = data['name']
+    arn = data.get('arn', None)
+    role = Role.query.filter_by(name=role_name).one_or_none()
+    if role is None:
+        if arn is None:
+            raise ApiException(
+                requests.codes.bad_request,
+                "Creating a role requires a role ARN"
+                )
+        if not VALID_NAME_RE.match(role_name):
+            raise ApiException(
+                requests.codes.bad_request,
+                "Invalid name for role"
+                )
+        role = Role(
+            id=generate_uuid(),
+            name=role_name,
+            arn=arn
+            )
+        db.session.add(role)
+        db.session.commit()
+        return _role_dict(role)
+    else:
+        raise ApiException(
+            requests.codes.conflict,
+            "Role name already exists"
+            )
+
+@app.route('/api/roles/<role_id>', methods=['PUT'])
+@api(enabled=ENABLE_USER_ENDPOINTS, require_admin=True, schema=ROLE_DETAILS_SCHEMA)
+@as_json
+def edit_role(role_id):
+    """
+    Edits a role.
+
+    The body of the request should contain a JSON object.
+    There is body should contain the parameters:
+        name(string): name of the role to operate on
+        arn(string): ARN of the IAM role associated with the Quilt role.
+
+    To change the name of a role, provide the new name you want to use along with
+        the current ARN.
+    To change the ARN attached to a role, provide the current name along with
+        the new ARN you want to use.
+    """
+    data = request.get_json()
+    role_name = data['name']
+    arn = data['arn']
+    role = Role.query.get(role_id)
+
+    # edit existing role
+    if not VALID_NAME_RE.match(role_name):
+        raise ApiException(
+            requests.codes.bad_request,
+            "Invalid name for role"
+            )
+    role.name = role_name
+    role.arn = arn
+    db.session.add(role)
+    db.session.commit()
+    return _role_dict(role)
+
+@app.route('/api/roles/<role_id>', methods=['DELETE'])
+@api(enabled=ENABLE_USER_ENDPOINTS, require_admin=True, schema=ROLE_DETAILS_SCHEMA)
+@as_json
+def delete_role(role_id):
+    """
+    Delete the role at role_id and remove it from associated user accounts.
+    """
+    # delete role
+    # must remove role from all users with that role due to foreign key constraint
+    role = Role.query.get(role_id)
+    users = User.query.filter_by(role_id=role.id).all()
```

Comment:
Using update will be much more efficient here. Although we don't expect millions of users in a single role, it's faster and cleaner to let the database do the work.

Reply:
Done